plugins/hls-notes-plugin/src/Ide/Plugin/Notes.hs

```diff
@@ -0,0 +1,125 @@
+module Ide.Plugin.Notes (descriptor, Log) where
+
+import           Control.Lens                     (ix, (^.), (^?))
+import           Control.Monad.Except
+import           Control.Monad.IO.Class           (liftIO)
+import           Control.Monad.Trans              (lift)
+import qualified Data.Array                       as A
+import           Data.HashMap.Strict              (HashMap)
+import qualified Data.HashMap.Strict              as HM
+import qualified Data.HashSet                     as HS
+import           Data.Maybe                       (catMaybes, listToMaybe,
+                                                   mapMaybe)
+import           Data.Text                        (Text, intercalate)
+import qualified Data.Text                        as T
+import qualified Data.Text.Utf16.Rope             as Rope
+import           Development.IDE                  hiding (line)
+import           Development.IDE.Core.PluginUtils (runActionE, useE)
+import           Development.IDE.Core.Shake       (toKnownFiles)
+import qualified Development.IDE.Core.Shake       as Shake
+import           Development.IDE.Graph.Classes    (Hashable, NFData)
+import           GHC.Generics                     (Generic)
+import           Ide.Plugin.Error                 (PluginError (..))
+import           Ide.Types
+import qualified Language.LSP.Protocol.Lens       as L
+import           Language.LSP.Protocol.Message    (Method (Method_TextDocumentDefinition),
+                                                   SMethod (SMethod_TextDocumentDefinition))
+import           Language.LSP.Protocol.Types
+import qualified Language.LSP.Server              as LSP
+import           Language.LSP.VFS                 (VirtualFile (..))
+import           Text.Regex.TDFA                  (Regex, caseSensitive,
+                                                   defaultCompOpt,
+                                                   defaultExecOpt,
+                                                   makeRegexOpts, matchAllText)
+
+data Log
+    = LogShake Shake.Log
+    | LogNotesFound NormalizedFilePath [(Text, Position)]
+    deriving Show
+
+data GetNotesInFile = MkGetNotesInFile
+    deriving (Show, Generic, Eq, Ord)
+    deriving anyclass (Hashable, NFData)
+type instance RuleResult GetNotesInFile = HM.HashMap Text Position
+
+data GetNotes = MkGetNotes
+    deriving (Show, Generic, Eq, Ord)
+    deriving anyclass (Hashable, NFData)
+type instance RuleResult GetNotes = HashMap Text (NormalizedFilePath, Position)
+
+instance Pretty Log where
+    pretty = \case
+        LogShake l -> pretty l
+        LogNotesFound file notes ->
+            "Found notes in " <> pretty (show file) <> ": ["
+            <> pretty (intercalate ", " (fmap (\(s, p) -> "\"" <> s <> "\" at " <> T.pack (show p)) notes)) <> "]"
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId = (defaultPluginDescriptor plId)
```

Comment:
Could we have some high-level documentation on how this plugin works? 
Something of the like "whenever a session is initialised, we run on all files of the component an index step for the notes available. On Goto Definition request, we look up the position"